### ISSUE TITLE
TODO list to support AllReduce

### ISSUE THREAD

**Issue by workingloong** [sql-machine-learning/elasticdl#2185]
Todo list to implement elastic AllReduce.

## Study the implementation of AllReduce in the existing libraries.
1. Study the implementation of elastic AllReduce in FTlib
2. Study AllReduce of PyTorch and Horovod.
3. Summarize the above implementations and choose a library to implement elastic AllReduce in ElasticDL.

## Develop AllReduce in ElasticDL
1. Master assigns ranks of worker pods to construct a collective communicator.
2. Master re-assigns ranks if some worker pods fail.
3. Use a communicator of a library to average gradients and update the model.
3. Evaluate the validation data set during the training per some steps.
4. Export model after the training.

## Benchmark for elastic AllReduce.
1. Prepare dataset like [ImageNet](http://image-net.org/download).
2. Choose a CV model like [MobileNetV3](https://github.com/Bisonai/mobilenetv3-tensorflow) to do an experiment on a distributed GPU cluster.
3. Test the performance using Horovod without elastic scheduling and ElasticDL AllReduce.

**Comment by skydoorkai** [comment]
# Allreduce Work Plan

## Write Design Doc
7/20 - 7/23 @workingloong  @skydoorkai 

There are two main parts
1. Consensus to get a group of workers to create/recreate a process group for allreduce.
2. Given a process group, support broadcast and allreduce.

We will use the ElasticDL master for consensus.
The master will monitor worker pods, get all running workers' IPs, assign ranks. Workers get these data from the master to create/recreate a process group for allreduce when needed. 

For data communications (broadcast, allreduce), we will use PyTorch distribute APIs similar to TFlib as the first implementation.

To support evaluation during training, we need to support regrouping a subset of workers as a process group, so that when some workers are doing the evaluation, other workers can continue training with allreduce.

## Benchmark
7/20 - 7/27 Benchmark ElasticDL Allreduce with FTlib @workingloong 

8/13 - 8/17 Benchmark ElasticDL Allreduce with our implementation @workingloong 

##  Development
7/27 - 8/12

### Consensus
7/27 - 8/5   @workingloong  @skydoorkai 

### Communications
8/3 - 8/6 Using PyTorch distribute API.  @workingloong

### Support evaluation
8/6 - 8/12   @workingloong

## Research

Better Allreduce implementation besides PyTorch distributed APIs. Can we use NCCL directly? How to support regroup after failure for NCCL? Is PS-based allreduce a better choice for fault-tolerant? @workingloong  @skydoorkai

**Comment by weberxie** [comment]
Any updates on this job? I'd like to contribute on this.

**Comment by workingloong** [comment]
> Any updates on this job? I'd like to contribute on this.

Sorry for the late reply. We have implemented elastic AllReduce using Horovod and you can see the design https://github.com/sql-machine-learning/elasticdl/blob/develop/docs/designs/allreduce_based_horovod.md.
